src/src/com/microsoft/aad/adal/IdToken.java

```diff
@@ -23,28 +23,145 @@
+        return mTenantId;
+    }
+
+    public String getUpn() {
+        return mUpn;
+    }
 
-    String mUpn;
+    public String getGivenName() {
+        return mGivenName;
+    }
+
+    public String getFamilyName() {
+        return mFamilyName;
+    }
 
-    String mGivenName;
+    public String getEmail() {
+        return mEmail;
+    }
+
+    public String getIdentityProvider() {
+        return mIdentityProvider;
+    }
+
+    public String getObjectId() {
+        return mObjectId;
+    }
+
+    public long getPasswordExpiration() {
+        return mPasswordExpiration;
+    }
+
+    public String getPasswordChangeUrl() {
+        return mPasswordChangeUrl;
+    }
+
+    private HashMap<String, String> parseJWT(String idtoken) throws AuthenticationException {
+        String idbody = extractJWTBody(idtoken);
+        // URL_SAFE: Encoder/decoder flag bit to use
+        // "URL and filename safe" variant of Base64
+        // (see RFC 3548 section 4) where - and _ are used in place of +
+        // and /.
+        byte[] data = Base64.decode(idbody, Base64.URL_SAFE);
+        HashMap<String, String> responseItems = new HashMap<String, String>();
+
+        try {
+            String decodedBody = new String(data, "UTF-8");
+            extractJsonObjects(responseItems, decodedBody);
+            return responseItems;
+        } catch (UnsupportedEncodingException exception) {
+            Logger.e(TAG, "The encoding is not supported.", "", ADALError.ENCODING_IS_NOT_SUPPORTED, exception);
+            throw new AuthenticationException(ADALError.ENCODING_IS_NOT_SUPPORTED, exception.getMessage(), exception);
+        } catch (JSONException exception) {
+            Logger.e(TAG, "Failed to parse the decoded body into JsonObject.", "", ADALError.JSON_PARSE_ERROR,
+                    exception);
+            throw new AuthenticationException(ADALError.JSON_PARSE_ERROR, exception.getMessage(), exception);
+        }
+    }
+
+    private String extractJWTBody(String idtoken) throws AuthenticationException {
+        int firstDot = idtoken.indexOf(".");
+        int secondDot = idtoken.indexOf(".", firstDot + 1);
+        int invalidDot = idtoken.indexOf(".", secondDot + 1);
+
+        if (invalidDot == -1 && firstDot > 0 && secondDot > 0) {
+            return idtoken.substring(firstDot + 1, secondDot);
+        } else {
+            throw new AuthenticationException(ADALError.IDTOKEN_PARSING_FAILURE, "Failed to extract the ClientID");
+        }
+    }
 
-    String mFamilyName;
+    private static void extractJsonObjects(HashMap<String, String> responseItems, String jsonStr) throws JSONException {
+        final JSONObject jsonObject = new JSONObject(jsonStr);
 
-    String mEmail;
+        final Iterator<?> i = jsonObject.keys();
 
-    String mIdentityProvider;
-    
-    String mObjectId;
-    
-    long mPasswordExpiration;
-    
-    String mPasswordChangeUrl;
-        
-    public IdToken(){
-        
+        while (i.hasNext()) {
```

Comment:
That's strange. Why do you need boxing in here and manual while loop? This should just work:
for(String key :  jsonObject.keys()) {
responseItems.put(key, jsonObject.getString(key));
}